File: homeworks/homework2/22002105_Gokmen_AhmetBerke.lua
function PrintArr(arr) for k, e in pairs(arr) do io.write(k) io.write(": ") io.write(e) io.write(" ") end io.write("\n") end

-- 1. Are formal and actial parameters type checked?
print("\n 1.")

function printElems(elem1, elem2) print(elem1) print(elem2) end
printElems("sa", 1)
printElems(1, "as")

-- 2. Are keyword (named) parameters supported?
print("\n 2.")

-- They are actually supported, buy its by using tables in lua
function kwParameters(tab)
  print(tab.arg1)
  print(tab.arg2)
end

kwParameters{arg1="berke", arg2="gokmen"} -- this is how, but this is actually a shorter syntax of passing a table to the function 
kwParameters({arg1="gokmen", arg2="berke"})

-- 3. Are default paremeters supported?
print("\n 3.")

-- They are supported, but not directly in the function parameter list
function defParameters(req, opt1, opt2) 
  opt1 = opt1 or 1 -- this is how we make default parameters
  opt2 = opt2 or 2
  print(req, opt1, opt2)
end

defParameters(10)
defParameters(10, 20)
defParameters(10, 20, 30)

-- 4. What are the paremeter passing methods provided?
print("\n 4.")

val = 10
list = {1, 2, 3}
dictionary = {a = 5, b = 10}

print("before", val)
PrintArr(list)
PrintArr(dictionary)

function testPassing(_val, _list, _dictionary)
  _val = _val + 10 -- does not change the actual value
  _list[#_list + 1] = 155 -- modifies the actual value
  _dictionary["new"] = 333 -- modifies the actual value
  print("inside", _val)
  PrintArr(_list)
  PrintArr(_dictionary)
end

testPassing(val, list, dictionary)
print("after", val)
PrintArr(list)
PrintArr(dictionary)

-- 5. Can subprograms be passed as parameters? If so, how is the referencing environment of the passed subprogram bound?

-- it seems that the referencing environment is shallow binded
print("\n 5.")

text = "outside"

function sendMessage(messageGenerator)
  text = "inside"
  print(messageGenerator())
end

function buildHiMessage() return "Hi " .. text end

sendMessage(buildHiMessage)

File: homeworks/homework2/22002105_Gokmen_AhmetBerke.rs
fn main() {
    // 1. Are formal and actial parameters type checked?
    println!("\n 1.");

    fn print_int(value: i64) {
        println!("Value: {value}");
    }

    print_int(10);
    // print_int("sa"); // Error 1, this will not compile

    // 2. Are keyword (named) parameters supported?
    println!("\n 2."); // They are not supported in rust

    fn kw_test(val1: i64, val2: i64) {
        println!("val1: {val1}, val2: {val2}");
    }

    kw_test(10, 10);
    // These will all fail (Error 2)
    // kw_test(val1=10, val2=20);
    // kw_test({val1=10, val2=20});
    // kw_test({val1: 10, val2: 20});
    // kw_test({val1 => 10, val2 => 20});

    // 3. Are default paremeters supported?
    println!("\n 3.");

    // Rust actually does not support default arguments.
    // The following function will generate a compile time error.
    // fn say_info(experience: i32 = 10, age: i32 = 20) {
    //     println!("hi! {experience} {age}");
    // }

    // 4. What are the paremeter passing methods provided?
    println!("\n 4.");

    // 4.1. Transferring ownership (similar to pass by value)
    fn transfer_ownership(name: String) {
        println!("Hi {}", name);
    }

    let name: String = String::from("berke gokmen");
    transfer_ownership(name);
    // println!("Outside: {}", name); // This will cause a compile time error

    // 4.2. Pass by reference (borrowing)
    fn borrow_value(name: &String) {
        println!("Hi again {}", name);
    }

    let animal: String = String::from("animal");
    borrow_value(&animal);
    println!("Outside again: {}", animal); // This will work since object is not transfered to other function

    // 4.2. Pass by reference (borrowing)
    fn borrow_and_modify(name: &mut String) {
        name.push_str("_modified");
        println!("Modified: {}", name)
    }

    let mut modify_name = String::from("selam");
    borrow_and_modify(&mut modify_name);
    println!("outside: {}", modify_name); // we see that variable changed

    // 5. Can subprograms be passed as parameters? If so, how is the referencing environment of the passed subprogram bound?
    println!("\n 5.");

    // Rust will choose this num (static scope). Thus it's deep binding.
    let num: i32 = 11;
    let get_num = || {
        println!("{}", num);
    };

    fn test_ref_env<F>(op: F)
    where
        F: Fn(),
    {
        let num: i32 = 22; // this will be ignored

        op();
    }

    test_ref_env(get_num);
}
